lib/Monotone_framework.ml

```diff
@@ -0,0 +1,958 @@
+             |Mir.Assignment ({texpr= Indexed ({texpr= Var s; _}, _); _}, _) ->
+                Map.remove m s
+            | Mir.Assignment (_, _) -> Errors.fatal_error ()
+            | Mir.TargetPE _
+             |Mir.NRFunApp (_, _)
+             |Mir.Check _ | Mir.Break | Mir.Continue | Mir.Return _ | Mir.Skip
+             |Mir.IfElse (_, _, _)
+             |Mir.While (_, _)
+             |Mir.For _ | Mir.Block _ | Mir.SList _ | Mir.FunDef _ ->
+                m )
+  end
+  : TRANSFER_FUNCTION
+    with type labels = int
+     and type properties = (string, Mir.expr_typed_located) Map.Poly.t option
+  )
+
+(** The transfer function for a copy propagation analysis *)
+let copy_propagation_transfer
+    (flowgraph_to_mir : (int, Mir.stmt_loc_num) Map.Poly.t) =
+  ( module struct
+    type labels = int
+    type properties = (string, Mir.expr_typed_located) Map.Poly.t option
+
+    let transfer_function l p =
+      match p with
+      | None -> None
+      | Some m ->
+          let mir_node = (Map.find_exn flowgraph_to_mir l).stmtn in
+          Some
+            ( match mir_node with
+            | Mir.Assignment
+                ( {texpr= Mir.Var s; _}
+                , {texpr= Mir.Var t; texpr_type; texpr_loc; texpr_adlevel} ) ->
+                Map.set m ~key:s
+                  ~data:
+                    {Mir.texpr= Mir.Var t; texpr_type; texpr_loc; texpr_adlevel}
+            | Mir.Decl {decl_id= s; _} | Mir.Assignment ({texpr= Var s; _}, _)
+              ->
+                Map.remove m s
+            | Mir.Assignment (_, _)
+             |Mir.TargetPE _
+             |Mir.NRFunApp (_, _)
+             |Mir.Check _ | Mir.Break | Mir.Continue | Mir.Return _ | Mir.Skip
+             |Mir.IfElse (_, _, _)
+             |Mir.While (_, _)
+             |Mir.For _ | Mir.Block _ | Mir.SList _ | Mir.FunDef _ ->
+                m )
+  end
+  : TRANSFER_FUNCTION
+    with type labels = int
+     and type properties = (string, Mir.expr_typed_located) Map.Poly.t option
+  )
+
+(** A helper function for building transfer functions from gen and kill sets *)
+let transfer_gen_kill p gen kill = Set.union gen (Set.diff p kill)
+
+(* TODO: from here *)
+
+(** Calculate the set of variables that a statement can assign to or declare *)
+let assigned_or_declared_vars_stmt
+    (s : (Mir.expr_typed_located, 'a) Mir.statement) =
+  match s with
+  | Mir.Assignment ({texpr= Var x; _}, _)
+   |Mir.Assignment ({texpr= Indexed ({texpr= Var x; _}, _); _}, _)
+   |Mir.Decl {decl_id= x; _}
+   |Mir.FunDef {fdname= x; _} ->
+      Set.Poly.singleton x
+  | Mir.Assignment (_, _) -> Errors.fatal_error ()
+  | Mir.TargetPE _ -> Set.Poly.singleton "target"
+  | Mir.NRFunApp (s, _) when String.suffix s 3 = "_lp" ->
+      Set.Poly.singleton "target"
+  | Mir.NRFunApp (_, _)
+   |Mir.Check _ | Mir.Break | Mir.Continue | Mir.Return _ | Mir.Skip
+   |Mir.IfElse (_, _, _)
+   |Mir.While (_, _)
+   |Mir.For _ | Mir.Block _ | Mir.SList _ ->
```

Comment:
For should declare and assign the loop variable, right?

Reply:
Oh whoops! Thanks for spotting that. Will fix it now.

Reply:
I added it so that my tests would work, but I think it should also be in the kill set like an assignment, which I didn't do

Reply:
Ah, I see you already fixed it. Sweet!